### PR TITLE
Increase sync timeouts

### PR DESCRIPTION
**Pull request by teor2345**
Increase the sync timeouts, to support slower mainnet and testnet nodes.

This PR is based on #980, because it modifies nearby code.

These longer timeouts are not triggered very often, because of the fixes in:
* #980: Ignore sync errors when the block is already verified
* #978: Make tips independent of response order
* #991: Always drop the final hash in peer responses
* #992: If the first ExtendTips hash is bad, discard it and re-check
* e5a6aa4c in this PR: Retry obtain and extend tips on failure

We might be able to reduce these timeouts in future, once we implement a peer reputation service.

In the meantime, users can further reduce the timeout rate by configuring the number of peers:
* fewer peers if the node is network-bound
* fewer peers to reduce the number of peers providing incorrect responses
* more peers if the node has too many slow peers
* more peers if verification is timing out before all the blocks are downloaded

As a side-effect, the longer timeouts in this PR also reduce the load on the network from zebrad instances which constantly fail to sync. (Because their network connections are slow or unreliable.)

TODO:
- [x] get PR #980 merged
- [x] rebase on main

### CODE REVIEW

**Comment by teor2345** [comment]
The clippy errors are unrelated to this PR, see #994.

File: zebra-network/src/constants.rs
Comment (hdevalence) [inline zebra-network/src/constants.rs]: Sorry, what's the connection between the buffer size and the block download timeout?

File: zebra-network/src/constants.rs
Comment (hdevalence) [inline zebra-network/src/constants.rs:86]: :+1:

File: zebra-network/src/constants.rs
Comment (hdevalence) [inline zebra-network/src/constants.rs:95]: :+1:

File: zebrad/src/commands/start/sync.rs
Comment (hdevalence) [inline zebrad/src/commands/start/sync.rs]: Can this just be hoisted out of the loop?

File: zebra-network/src/constants.rs
Comment (hdevalence) [reply zebra-network/src/constants.rs]: I mean, the number 10 for the peer set buffer size was made up totally artificially (it should probably be at least 2 or 3 to avoid sender contention, but otherwise there's no particular reason for it to be 10), so it's strange to see it justified in this way.

File: zebrad/src/commands/start/sync.rs
Comment (teor2345) [reply zebrad/src/commands/start/sync.rs]: Sure, we could check the loop condition twice - once for the logging, and once for the loop.

It doesn't make sense to log "waiting for pending blocks" if we're not actually going to wait.

File: zebra-network/src/constants.rs
Comment (teor2345) [reply zebra-network/src/constants.rs]: > what's the connection between the buffer size and the block download timeout?

The timeout needs to be approximately the amount of time it takes to drain a full buffer, and one pending block request per peer, via the slowest network speed we support.

Otherwise, we can choke the local network connection with block downloads, leading to repeated failures.

I'll revise the comment and decrease the buffer size - I think this comment might actually belong in sync.rs.

File: zebrad/src/commands/start/sync.rs
Comment (teor2345) [reply zebrad/src/commands/start/sync.rs]: Fixed in 53a2fcdc9e8f02bf3953014ba729a8937c75243c

File: zebra-network/src/constants.rs
Comment (teor2345) [reply zebra-network/src/constants.rs]: Fixed in 31e718a0ebad8ad32ae804790c54ba7b08e7d95d.

I decreased the buffer size to 3.

I also updated the sync timeout to 100, to match the default peer set size of 50, on a 10 Mbps node, downloading maximum-size blocks. (In practice, the actual blocks are smaller, but the calculation doesn't include latency. So I think they should balance out. Resets should also be rare, after this PR merges.)

I added some additional comments and documentation for network-constrained nodes.

File: zebrad/src/commands/start/sync.rs
Comment (hdevalence) [inline zebrad/src/commands/start/sync.rs:194]: Non-blocking on this PR, but when we restart the sync process, we retry `obtain_tips`.  Is there a benefit to special-casing this retry logic here?

**Review by hdevalence — APPROVED** [review]
This looks really good, particularly the rationale about the new EWMA parameters and timeouts is much better than the previous values, which were just made up.  Thanks for working through all of the edge cases here.

File: zebrad/src/commands/start/sync.rs
Comment (teor2345) [reply zebrad/src/commands/start/sync.rs:194]: I don't think there is much of a benefit in the current code.

The TIPS_RETRY_TIMEOUT is shorter, but not by much. So always using the SYNC_RESTART_TIMEOUT would be simpler and more reliable.

I think it's a holdover from when the gap between the TIPS_RETRY_TIMEOUT and SYNC_RESTART_TIMEOUT was much larger.

Comment (teor2345) [reply zebrad/src/commands/start/sync.rs:194]: See #1033 for a fix for this issue.

**Pull request merged**
